Enable creating new sketches from a template directory

gsingh93 opened the pull request:
## Please check if the PR fulfills these requirements

See [how to contribute](https://arduino.github.io/arduino-cli/latest/CONTRIBUTING/)

- [X] The PR has no duplicates (please search among the [Pull Requests](https://github.com/arduino/arduino-cli/pulls)
      before creating one)
- [X] The PR follows
      [our contributing guidelines](https://arduino.github.io/arduino-cli/latest/CONTRIBUTING/#pull-requests)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [ ] `UPGRADING.md` has been updated with a migration guide (for breaking changes)
- [ ] `configuration.schema.json` updated if new parameters are added.

## What kind of change does this PR introduce?

Implements feature request in #2239 

## What is the current behavior?

Running `arduino-cli sketch new MySketch` will create a `MySketch/MySketch.ino` file.

## What is the new behavior?

If `directories.template` is not set, the behavior is the same as before.

If it is set, running `arduino-cli sketch new MySketch` will copy the `directories.template` directory to create the `MySketch` directory.

## Does this PR introduce a breaking change, and is [titled accordingly](https://arduino.github.io/arduino-cli/latest/CONTRIBUTING/#breaking)?

This isn't a breaking change.

## Other Information

Tests/docs haven't been written. Putting this up as a draft to get initial feedback before putting in more effort. One question how to handle setting `sketchMainFilePath`. Should we rename some file in the template directory to `MySketch.ino`? Or just pick a .ino from the directory (ideally there's only one) and return that from the RPC call?

cmaglie commented on commands/sketch/new.go:
```diff
@@ -55,19 +55,32 @@ func NewSketch(ctx context.Context, req *rpc.NewSketchRequest) (*rpc.NewSketchRe
 		return nil, err
 	}
 
+	templateDir := configuration.Settings.GetString("directories.template")
```
I would not use `directories.template`, for two reasons:
* the prefix "directory." refers to places were there may be multiple things inside, but the sketch template is only one
* "template" is a bit too generic

IMHO is better `sketch.template`.

cmaglie commented on commands/sketch/new.go:
```diff
@@ -55,19 +55,32 @@ func NewSketch(ctx context.Context, req *rpc.NewSketchRequest) (*rpc.NewSketchRe
+			return nil, &arduino.CantCreateSketchError{Cause: err}
+		}
+		// TODO: Make this customizable?
+		sketchMainFilePath = sketchDirPath.Join("main.ino")
```
This won't work, `main.ino` must be named the same as the sketch folder (`/path/to/my_sketch/my_sketch.ino` for example).
Since the sketch folder is given as a command line argument by the user, you should do something similar to the `else` side of the `if`:
```go
sketchName := sketchDirPath.Base()
sketchMainFilePath = sketchDirPath.Join(sketchName + globals.MainFileValidExtension)
```
probably these two lines can be factored out of the if.

cmaglie reviewed (commented):
> One question how to handle setting sketchMainFilePath. Should we rename some file in the template directory to MySketch.ino? Or just pick a .ino from the directory (ideally there's only one) and return that from the RPC call?

mmmhh... It's a bit more convoluted than expected, we should consider these factors:

1. A sketch may be composed of many files
2. The **main** sketch file is the .ino file that is named exactly as the containing folder
3. We should copy the entire "template" sketch folder into the new user-requested sketch folder
4. In the new folder we must rename the template-main-sketch file to match the new folder name.

So let's recap with an example:

Suppose we configure the sketch template to point to `/path/to/sketch/template`, that has the following content:

```
/path/to/sketch/template/another_file.ino
/path/to/sketch/template/template.ino      <--- MAIN FILE
/path/to/sketch/template/passwords.h
```

if we run `arduino-cli sketch new mysketchbook/newsketch`, the expected result is:

```
./mysketchbook/newsketch/another_file.ino
./mysketchbook/newsketch/newsketch.ino      <--- MAIN FILE
./mysketchbook/newsketch/passwords.h
```

The most notable part here is that `template.ino` has been renamed `newsketch.ino` because it's the main file.

gsingh93 commented on configuration/defaults.go:
```diff
@@ -69,4 +69,6 @@ func SetDefaults(settings *viper.Viper) {
 	settings.BindEnv("directories.Data", "ARDUINO_DATA_DIR")
 	settings.BindEnv("sketch.always_export_binaries", "ARDUINO_SKETCH_ALWAYS_EXPORT_BINARIES")
+
+	settings.SetDefault("sketch.template", "")
```
This didn't fit with the other settings, not sure if I should create a section here like `// misc sketch settings`.

cmaglie replied:
Nope just keep it there, we will adjust the ordering in the future if more settings will come up.

gsingh93 commented:
@cmaglie thanks for the feedback! I've addressed your comments and rebased. If the template directory is `/path/to/template`, it's now required that `/path/to/template.ino` exists, and it will be renamed to the actual main sketch file.

If this general approach looks correct to you, I'll start working on the tests/docs when I get some time.

cmaglie commented:
Looks good, thanks, go on!

umbynos commented:
Hello @gsingh93 did you had time to continue? :smile:

gsingh93 commented:
Sorry, I haven't had time to work on this in a while. Someone else can pick it up from here, or it could be closed for now and I'll reopen if I ever come back to it. Either way is fine.
